File: src/robflex_api.c
#define _POSIX_C_SOURCE 200809L
#include <robflex_api.h>
#include <cJSON.h>
#include <stdarg.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <assert.h>
#include <stdatomic.h>
#include <sys/time.h>
#include <time.h>

__thread int socket_fd = 0;
int send_command_to_daemon(const char *cmd_json) {
    if (socket_fd == 0) {
        socket_fd = socket(AF_UNIX, SOCK_DGRAM, 0);
        if (socket_fd < 0) {
            perror("socket");
            return -1;
        }
    }

    struct sockaddr_un server_addr;
    memset(&server_addr, 0, sizeof(struct sockaddr_un));
    server_addr.sun_family = AF_UNIX;
    strncpy(server_addr.sun_path, SOCKET_PATH, sizeof(server_addr.sun_path) - 1);

    ssize_t num_bytes = sendto(socket_fd, cmd_json, strlen(cmd_json), 0,
                               (struct sockaddr *)&server_addr, sizeof(struct sockaddr_un));
    if (num_bytes < 0) {
        // perror("sendto");
        return -1;
    }
    return 0;
}

int robflex_set_scheduler(pid_t tid, int policy, int priority) {
    char cmd_json[MAX_JSON_SIZE];

    cJSON *root = cJSON_CreateObject();
    cJSON_AddStringToObject(root, "cmd", "sched_setscheduler");
    cJSON_AddNumberToObject(root, "tid", tid? tid : gettid());
    cJSON_AddNumberToObject(root, "policy", policy);
    cJSON_AddNumberToObject(root, "priority", priority);

    char *cmd_str = cJSON_PrintUnformatted(root);
    int result = send_command_to_daemon(cmd_str);

    cJSON_Delete(root);
    free(cmd_str);

    return result;
}

int robflex_set_priority(pid_t tid, int priority) {
    char cmd_json[MAX_JSON_SIZE];

    cJSON *root = cJSON_CreateObject();
    cJSON_AddStringToObject(root, "cmd", "set_priority");
    cJSON_AddNumberToObject(root, "tid", tid? tid : gettid());
    cJSON_AddNumberToObject(root, "priority", priority);

    char *cmd_str = cJSON_PrintUnformatted(root);
    int result = send_command_to_daemon(cmd_str);

    cJSON_Delete(root);
    free(cmd_str);

    return result;
}

int robflex_log_message(pid_t tid, const char *message, ...) {
    char cmd_json[MAX_JSON_SIZE];
    char formatted_message[MAX_LOG_SIZE];
    int result = -1;

    if(!tid){
        tid = gettid();
    }

    va_list args;
    va_start(args, message);
    vsnprintf(formatted_message, sizeof(formatted_message), message, args);
    va_end(args);

    cJSON *root = cJSON_CreateObject();
    cJSON_AddStringToObject(root, "cmd", "log_message");
    cJSON_AddNumberToObject(root, "tid", tid);
    cJSON_AddStringToObject(root, "message", formatted_message);

    char *cmd_str = cJSON_PrintUnformatted(root);
    result = send_command_to_daemon(cmd_str);

    cJSON_Delete(root);
    free(cmd_str);

    return result;
}

int robflex_update_ctrl_time_cost(pid_t tid, int value_in_us) {
    char cmd_json[MAX_JSON_SIZE];
    int result = -1;

    cJSON *root = cJSON_CreateObject();
    cJSON_AddStringToObject(root, "cmd", "update_ctrl_time_cost");
    cJSON_AddNumberToObject(root, "tid", tid? tid : gettid());
    cJSON_AddNumberToObject(root, "value_in_us", value_in_us);

    char *cmd_str = cJSON_PrintUnformatted(root);
    result = send_command_to_daemon(cmd_str);

    cJSON_Delete(root);
    free(cmd_str);

    return result;
}

static SystemData *ptr_shmem = NULL;
void* _init_shmem_data(const char *shmem_name){
    if(ptr_shmem == NULL){
        int fd = shm_open(shmem_name, O_RDONLY, 0);
        if(fd == -1){   
            perror("shm_open");
            return NULL;
        }
        ptr_shmem = (SystemData *)mmap(NULL, sizeof(SystemData), PROT_READ, MAP_SHARED, fd, 0);
        if(ptr_shmem == MAP_FAILED){
            perror("mmap");
            ptr_shmem = NULL;
        }
        close(fd);
    }
    return ptr_shmem;
}

enum SystemBusyDegree robflex_system_busy_degree() {
    assert(ptr_shmem != NULL);
    return atomic_load(&ptr_shmem->busy_degree);
}

static uint64_t robflex_get_time_ns() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + (uint64_t)ts.tv_nsec;
}

// assume signal handler is not running
int robflex_init_local_context(){
    loc_ctx.avg_timecost_ns = DEFUALT_PERIOD_TIME_IN_NS;
    loc_ctx.run_mode = PREDETERMINED;
    atomic_store_explicit(&loc_ctx.aux.norm.time_slice_ns, DEFUALT_PERIOD_TIME_IN_NS, memory_order_relaxed);
    loc_ctx.aux.norm.time_budgets = 0;
    atomic_store_explicit(&loc_ctx.in_critical, 0, memory_order_relaxed);
    atomic_store_explicit(&loc_ctx.n_signal_pendings, 0, memory_order_relaxed);
    return 0;
}

int robflex_set_cycles_for_tick(uint64_t cycles){
    atomic_store(&loc_ctx.in_critical, 1);
    if(loc_ctx.run_mode != YIELDING && loc_ctx.run_mode != PREDETERMINED){
        return -1;
    }
    // TODO: reset cycles_num for perf_fd
    
    loc_ctx.aux.norm.time_budgets = 0;
    atomic_store(&loc_ctx.in_critical, 0);
    while(atomic_exchange(&loc_ctx.n_signal_pendings, 0)){
        handle_tick();
    }
    return 0;
}

int robflex_set_time_for_throttle(uint64_t time_ns){
    atomic_store(&loc_ctx.in_critical, 1);
    if(loc_ctx.run_mode != YIELDING && loc_ctx.run_mode != PREDETERMINED){
        return -1;
    }
    loc_ctx.aux.norm.time_slice_ns = time_ns;
    loc_ctx.aux.norm.time_budgets = 0;
    atomic_store(&loc_ctx.in_critical, 0);
    while(atomic_exchange(&loc_ctx.n_signal_pendings, 0)){
        handle_tick();
    }
    return 0;
}

int robflex_set_as_immediate(uint64_t prio){
    loc_ctx.run_mode = IMMEDIATE;
    return 0;
}

int robflex_set_as_latency_flick(uint64_t lat_ns, uint64_t base_ns){
    atomic_store(&loc_ctx.in_critical, 1);
    loc_ctx.run_mode = LATENCY_ORIENTED;
    loc_ctx.aux.lat.target_lat = lat_ns;
    loc_ctx.aux.lat.start_time = robflex_get_time_ns();
    loc_ctx.aux.lat.hist_ncycle = base_ns;
    atomic_store(&loc_ctx.in_critical, 0);
    while(atomic_exchange(&loc_ctx.n_signal_pendings, 0)){
        handle_tick();
    }
    return 0;
}

int robflex_add_runcycle(uint64_t runcycle){
    atomic_store(&loc_ctx.in_critical, 1);
    loc_ctx.aux.lat.used_ncycle += runcycle;
    atomic_store(&loc_ctx.in_critical, 0);
    while(atomic_exchange(&loc_ctx.n_signal_pendings, 0)){
        handle_tick();
    }
    return 0;
}

int robflex_shot_on_latency(){
    atomic_store(&loc_ctx.in_critical, 1);
    if(loc_ctx.run_mode != LATENCY_ORIENTED){
        return -1;
    }
    uint64_t current_time = robflex_get_time_ns();
    uint64_t pass_time = current_time - loc_ctx.aux.lat.start_time;
    long long rest_time = (long long)loc_ctx.aux.lat.target_lat - (long long)pass_time;
    loc_ctx.aux.lat.hist_ncycle = loc_ctx.aux.lat.hist_ncycle*0.5 + loc_ctx.aux.lat.used_ncycle*0.5;
    // loc_ctx.aux.lat.start_time = current_time;
    loc_ctx.aux.lat.used_ncycle = 0;
    atomic_store(&loc_ctx.in_critical, 0);

    while(atomic_exchange(&loc_ctx.n_signal_pendings, 0)){
        handle_tick();
    }

    // printf("target latency: %llu, pass time: %llu, rest time: %lld, hist_ncycle: %llu, used_ncycle: %llu\n", loc_ctx.aux.lat.target_lat, pass_time, rest_time, loc_ctx.aux.lat.hist_ncycle, loc_ctx.aux.lat.used_ncycle);
    if(rest_time > 1000000) {
        struct timespec sleep_ts;
        sleep_ts.tv_sec = 0;
        sleep_ts.tv_nsec = rest_time - 1000000;
        robflex_log_message(gettid(), "rest_time:%llu, sleep_ns:%llu", rest_time, rest_time - 1000000);
        nanosleep(&sleep_ts, NULL);
    }

    loc_ctx.aux.lat.start_time = robflex_get_time_ns();
    return 0;
}

int robflex_switch_context(LocalContext *new_context, LocalContext *saved_context){
    LocalContext loc_ctx_backup;
    atomic_store(&loc_ctx.in_critical, 1);
    if(saved_context != NULL){
        loc_ctx_backup = loc_ctx;
    }
    loc_ctx = *new_context;
    atomic_store(&loc_ctx.in_critical, 0);
    while(atomic_exchange(&loc_ctx.n_signal_pendings, 0)){
        handle_tick();
    }
    // probably saved_context is new_context, so we need to copy the data from loc_ctx_backup to saved_context
    if(saved_context != NULL){
        *saved_context = loc_ctx_backup;
    }
    return 0;
}

// only for test
int robflex_switch_context_block(LocalContext *new_context, LocalContext *saved_context){
    LocalContext loc_ctx_backup;
    sigset_t set;
    sigemptyset(&set);
    sigaddset(&set, PERF_SIGNAL);

    sigprocmask(SIG_BLOCK, &set, NULL);
    if(saved_context != NULL){
        loc_ctx_backup = loc_ctx;
    }
    loc_ctx = *new_context;
    sigprocmask(SIG_UNBLOCK, &set, NULL);

    // probably saved_context is new_context, so we need to copy the data from loc_ctx_backup to saved_context
    if(saved_context != NULL){
        *saved_context = loc_ctx_backup;
    }    
    return 0;
}

File: src/python/robflex_api.py
import socket
import os
import json
import threading
from typing import Optional, Any

# ================= 配置区域 =================
# 请确保这里的 SOCKET_PATH 与 C 语言代码中的宏定义一致
# 例如: "/tmp/robflex_daemon.sock"
SOCKET_PATH = "/tmp/schedule_daemon.sock"

# 模拟 C 代码中的宏大小限制 (可选，Python 动态字符串通常不需要，但为了逻辑一致保留)
MAX_JSON_SIZE = 4096 
MAX_LOG_SIZE = 1024

# ================= 内部实现 =================

# 使用 thread-local 存储 socket fd，对应 C 语言的 __thread int socket_fd
_local = threading.local()

def _get_socket() -> socket.socket:
    """获取当前线程的 socket 连接，如果没有则创建 (对应 C 中的懒加载逻辑)"""
    if not hasattr(_local, 'sock') or _local.sock is None:
        try:
            # AF_UNIX, SOCK_DGRAM 对应 C 代码中的 socket(AF_UNIX, SOCK_DGRAM, 0)
            _local.sock = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
            # 设置超时时间，避免永久阻塞 (C 代码默认是阻塞的，Python 建议加个超时以防死锁)
            _local.sock.settimeout(5.0) 
        except OSError as e:
            raise RuntimeError(f"Failed to create socket: {e}")
    return _local.sock

def _send_command_to_daemon(cmd_dict: dict) -> int:
    """
    内部发送函数
    返回: 0 (成功), -1 (失败)
    """
    try:
        sock = _get_socket()
        
        # 序列化 JSON (对应 cJSON_PrintUnformatted)
        cmd_json = json.dumps(cmd_dict, separators=(',', ':'))
        
        if len(cmd_json) > MAX_JSON_SIZE:
            print(f"Warning: JSON payload too large ({len(cmd_json)} > {MAX_JSON_SIZE})")

        # 构建地址结构 (对应 struct sockaddr_un)
        # Python 的 sendto 对于 AF_UNIX 可以直接传路径字符串，或者 bytes 路径
        # 为了严格对应 C 的结构体行为，我们传入路径
        server_addr = SOCKET_PATH
        
        # 发送数据 (对应 sendto)
        # 注意：SOCK_DGRAM 是无连接的，每次 sendto 都需要指定地址
        num_bytes = sock.sendto(cmd_json.encode('utf-8'), server_addr)
        
        if num_bytes < 0:
            return -1
            
        return 0
        
    except Exception as e:
        # 对应 C 代码中注释掉的 perror，这里打印到 stderr 或日志
        # print(f"Send error: {e}") 
        return -1

# ================= 公开 API (对应 C 函数) =================

def robflex_set_scheduler(tid: int, policy: int, priority: int) -> int:
    """
    设置调度策略
    对应 C: int robflex_set_scheduler(pid_t tid, int policy, int priority)
    """
    if tid == 0:
        tid = os.gettid()
    
    payload = {
        "cmd": "sched_setscheduler",
        "tid": tid,
        "policy": policy,
        "priority": priority
    }
    
    return _send_command_to_daemon(payload)

def robflex_set_priority(tid: int, priority: int) -> int:
    """
    设置优先级
    对应 C: int robflex_set_priority(pid_t tid, int priority)
    """
    if tid == 0:
        tid = os.gettid()
        
    payload = {
        "cmd": "set_priority",
        "tid": tid,
        "priority": priority
    }
    
    return _send_command_to_daemon(payload)

def robflex_log_message(tid: int, message: str, *args: Any) -> int:
    """
    记录日志 (支持格式化字符串)
    对应 C: int robflex_log_message(pid_t tid, const char *message, ...)
    
    用法:
    robflex_log_message(0, "CPU usage is %d%%", 85)
    """
    # if tid == 0:
    #     tid = os.gettid()
    
    # 处理可变参数格式化 (对应 vsnprintf)
    if args:
        try:
            formatted_message = message % args
        except TypeError:
            # 如果用户传入的是 {} 格式而不是 % 格式，尝试 format
            formatted_message = message.format(*args)
    else:
        formatted_message = message

    if len(formatted_message) > MAX_LOG_SIZE:
        formatted_message = formatted_message[:MAX_LOG_SIZE]

    payload = {
        "cmd": "log_message",
        "tid": tid,
        "message": formatted_message
    }
    
    return _send_command_to_daemon(payload)

def robflex_update_ctrl_time_cost(tid: int, value_in_us: int) -> int:
    """
    更新控制时间成本
    对应 C: int robflex_update_ctrl_time_cost(pid_t tid, int value_in_us)
    """
    if tid == 0:
        tid = os.gettid()
        
    payload = {
        "cmd": "update_ctrl_time_cost",
        "tid": tid,
        "value_in_us": value_in_us
    }
    
    return _send_command_to_daemon(payload)

# ================= 清理函数 (可选) =================

def robflex_cleanup():
    """
    关闭当前线程的 socket。
    通常不需要手动调用，进程退出时会自动清理。
    如果在长生命周期程序中需要重置连接，可以调用此函数。
    """
    if hasattr(_local, 'sock') and _local.sock is not None:
        try:
            _local.sock.close()
        except:
            pass
        _local.sock = None

# ================= 使用示例 =================
if __name__ == "__main__":
    print(f"Connecting to daemon at: {SOCKET_PATH}")
    
    # 1. 测试日志
    ret = robflex_log_message(0, "Python test log: Hello from PID %d", 0)
    print(f"log_message result: {ret}")
    
    # # 2. 测试更新 CPI (假设值为 100 百万分之一)
    # ret = robflex_update_ctrl_time_cost(0, 100)
    # print(f"update_ctrl_time_cost result: {ret}")
    
    # # 3. 测试设置调度策略 (SCHED_FIFO = 1, Priority = 10)
    # # 注意：这需要守护进程有 CAP_SYS_NICE 权限才能实际生效
    # ret = robflex_set_scheduler(0, 1, 10)
    # print(f"set_scheduler result: {ret}")
    
    # # 4. 测试设置优先级
    # ret = robflex_set_priority(0, 20)
    # print(f"set_priority result: {ret}")
    
    # 清理
    robflex_cleanup()